crates/matrix-sdk/src/room/mod.rs

```diff
@@ -3201,103 +3201,104 @@ impl Room {
-            let mut missing_room_members: bool = false;
 
             loop {
                 // This is equivalent to a combine stream operation, triggering a new emission
                 // when any of the branches changes
                 tokio::select! {
-                    Some((next, _)) = requests_stream.next() => { new_event = Some(next); }
-                    Some(next) = seen_request_ids_stream.next() => { seen_ids = next; }
-                    Some(next) = room_info_stream.next() => {
-                        missing_room_members = !next.are_members_synced()
-                    }
-                    else => break,
-                }
-
-                // We need to emit new items when we may have missing room members:
-                // this usually happens after a gappy (limited) sync
-                let has_missing_room_members = prev_missing_room_members != missing_room_members;
-                if has_missing_room_members {
-                    prev_missing_room_members = missing_room_members;
-                }
-
-                // We need to emit new items if the seen join request ids have changed
-                let has_new_seen_ids = prev_seen_ids != seen_ids;
-                if has_new_seen_ids {
-                    prev_seen_ids = seen_ids.clone();
-                }
-
-                if let Some(SyncStateEvent::Original(event)) = new_event.clone() {
-                    // Reset the new event value so we can check this again in the next loop
-                    new_event = None;
-
-                    // If we can calculate the membership change, try to emit only when needed
-                    if event.prev_content().is_some() {
-                        match event.membership_change() {
-                            MembershipChange::Banned |
-                            MembershipChange::Knocked |
-                            MembershipChange::KnockAccepted |
-                            MembershipChange::KnockDenied |
-                            MembershipChange::KnockRetracted => {
-                                match this.clone().get_current_join_requests(&seen_ids).await {
+                    Some((event, _)) = requests_stream.next() => {
+                        if let Some(event) = event.as_original() {
+                            // If we can calculate the membership change, try to emit only when needed
+                            if event.prev_content().is_some() {
+                                match event.membership_change() {
+                                    MembershipChange::Banned |
+                                    MembershipChange::Knocked |
+                                    MembershipChange::KnockAccepted |
+                                    MembershipChange::KnockDenied |
+                                    MembershipChange::KnockRetracted => {
+                                        match this.get_current_join_requests(&seen_ids).await {
```

Comment:
Instead of duplicating the `match this.get_current_join_requests(,..`) (which ought to be renamed to s/join/knock/), could we have a bool `emit`, and then, `if emit { match this.get_current_join_request() { ... }}` to avoid the code duplication?